open-sphere-plugins/ogc-server/src/main/java/io/opensphere/server/control/AbstractServerSourceController.java

```diff
@@ -331,9 +349,69 @@ protected void updateSource(IDataSource source)
         finally
         {
             myConfigLock.writeLock().unlock();
+            if (myReloadListenerMap.containsKey(source))
+            {
+                EventListener<AbstractSingleStateEvent> listener = myReloadListenerMap.get(source);
+                myReloadListenerMap.remove(source);
+                listener.notify(new AbstractSingleStateEvent()
+                {
+                    @Override
+                    public String getDescription()
+                    {
+                        return "Deactivation of " + source.getName() + " is done, finishing reload.";
+                    }
+                });
+            }
         }
     }
 
+    /** Reload sources which are marked as active, but are not actually activated. */
+    protected void reloadBadSources()
+    {
+        Set<String> activeSourceURLs = getActiveSourceURLs();
+        getSourceList().stream().forEach(source ->
+        {
+            if (source instanceof URLSource)
+            {
+                if (source.isActive() && !activeSourceURLs.contains(((URLSource)source).getURLString()))
+                {
+                    myReloadListenerMap.put(source, new EventListener<AbstractSingleStateEvent>()
+                    {
+                        @Override
+                        public void notify(AbstractSingleStateEvent event)
+                        {
+                            activateSource(source);
+                        }
+                    });
+                    deactivateSource(source);
+                }
+            }
+        });
+    }
+
+    /**
+     * Gets the set of active sources.
+     *
+     * @return the set of active sources
+     */
+    private Set<String> getActiveSourceURLs()
+    {
+        Set<String> activeURLs = new HashSet<String>();
+        DataGroupController dgc = MantleToolboxUtils.getMantleToolbox(myToolbox).getDataGroupController();
+        dgc.getActiveMembers(false).stream().forEach(dti ->
+        {
+            if (UrlUtil.isValidAbsoluteUrl(dti.getUrl()))
+            {
+                activeURLs.add(dti.getUrl());
+            }
+            else if (UrlUtil.isValidAbsoluteUrl(dti.getSourcePrefix()))
```

Comment:
Can we fix the dti's that don't have getUrl instead of using source prefix.

Reply:
I fixed this for STK sources